README.md
<div id="top"></div>

<!-- BADGE -->
[![Release Date](https://img.shields.io/github/release-date/Piero24/AmzUrlGen)]()
[![Last Commit](https://img.shields.io/github/last-commit/Piero24/AmzUrlGen)]()
[![Maintenance](https://img.shields.io/badge/Maintained-yes-green.svg)](https://github.com/Piero24/AmzUrlGen)
<!--[![Maintenance](https://img.shields.io/badge/Maintained%3F-no-red.svg)](https://github.com/Piero24/AmzUrlGen) -->
[![Issues](https://img.shields.io/github/issues/Piero24/AmzUrlGen)]()
[![Issues Pr](https://img.shields.io/github/issues-pr/Piero24/AmzUrlGen)]()
[![Downloads](https://img.shields.io/github/downloads/Piero24/AmzUrlGen/total)]()

[![License](https://img.shields.io/github/license/Piero24/AmzUrlGen)]()
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](code_of_conduct.md)


<!--
*** You can make other badges here
*** [shields.io](https://shields.io/)
*** or here
*** [CircleCI](https://circleci.com/)
-->
<br/>

<!-- Working ✅ Work in progress ⚠️ ~~Deprecate~~ ⛔️ -->
**Project Status:** Working ✅ <a href="#maintenance">⇩</a>
<br/>
**Percentage Completely:** 70% 🔋 <a href="#roadmap">⇩</a>
<br/>
<br/>

<p align="center">
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/logo%20AmzMsgRRBot.png" width="100" height="100">
</p>

<h1 align="center">
    <a href="https://github.com/Piero24/AmzUrlGen">AmazonUrlGen</a>
</h1>

> *<p align="center">AmzUrlGen is a telegram bot that allows you to create URLs for your listings on Amazon.
  These links allow you to rank faster and better for the keywords you have chosen.
  So you can increase the sales of your products on the Amazon Marketplace.*
  
<p align="center">
    The Index for this project is available here
    <br/>
    <a href="#index"><strong>Explore the Index »</strong></a>
    <br/>
    <br/>
    🇮🇹  
    <a href="https://github.com/Piero24/AmzUrlGen/blob/main/README-IT.md"><strong>Italian</strong></a> 
    version
    <br/>
    <br/>
    <a href="https://github.com/Piero24/AmzUrlGen">View Demo</a>
    •
    <a href="https://github.com/Piero24/AmzUrlGen/issues">Report Bug</a>
    •
    <a href="https://github.com/Piero24/AmzUrlGen/issues">Request Feature</a>
  </p>



---



<h2 id="itroduction" align="center"><br/><br/><br/>📔  Itroduction  📔</h2>
          
<p align="center">
  Thanks to this script you can have your own personal telegram bot to create the links for the listings.
  <br/>
  The bot allows you to generate the best links to better index your products.
  <br/>
  These are the available links and their explanation:
  <br/>
  <br/>
  <br/>
  🧙‍ <strong>MAGIC LINK</strong> 🧙‍
  <br/>
  SuperURL that mimics external traffic. Proven to be the best ranking URL.
  <br/>
  <br/>
  🐲 <strong>SUPREME URL 2.0</strong> 🐲
  <br/>
  2-step URL that provides incredible ranking results. This URL restricts products via an ASIN and a keyword.
  <br/>
  Ranking choice # 1 After many tests, we have seen fantastic results with this URL.
  <br/>
  <br/>
  🐉 <strong>SUPREME URL</strong> 🐉
  <br/>
  This is the ultimate Super URL generator. Our proprietary software generates the EXACT 
  <br/>
  Amazon URL used to determine search time, keyword attribution and more. 
  <br/>
  Drive buyers to the product page, increasing conversions. 
  <br/>
  There is no other Super URL out there like it!
  <br/>
  <br/>
  🔎 <strong>SEARCH FIND BUY SEEKER URL (SFB)</strong> 🔎
  <br/>
  This URL will allow your customers to find the product in the search results. 
  <br/>
  We try to find the exact page where your ad is located.
  <br/>
  In some cases, your customers may need to search multiple pages.
  <br/>
  <br/>
  👟 <strong>2-STEP BRAND URL</strong> 👟
  <br/>
  This Super URL filters your required product by brand and keyword and 
  <br/>
  attributes keywords that are directly related to your product title. 
  <br/>
  This URL does not include the timestamp feature and directs buyers to click on the listing before purchasing.
  <br/>
  <br/>
  🌱 <strong>BUY TOGETHER URL</strong> 🌱
  <br/>
  This Super URL allows you to manipulate Amazon's "frequently bought together" section of 
  <br/>
  your listing, driving traffic to a button that adds two items to your cart. 
  <br/>
  With enough purchases of two items, they will appear in the "Frequently Purchased Together" section for both listings.
  <br/>
  <br/>
  🛒 <strong>ADD TO CART URL</strong> 🛒
  <br/>
  Conversions generated via this URL give Amazon the impression that the purchase originated from Amazon Alexa. 
  <br/>
  With enough purchases, this increases the likelihood of earning the 
  <br/>
  "Amazon Choice" badge, often leading to increased organic sales.
  <br/>
  <br/>
  🔑 <strong>CANONICAL URL</strong> 🔑
  <br/>
  25% of Amazon's search traffic comes from search engines, 
  <br/>
  so creating this link allows you to optimize your page for search engine traffic outside of Amazon.
  <br/>
  Target 5 keywords with this URL and share on PBN or blog with anchor link.
  <br/>
  So, watch it climb the Google rankings, based on your keywords!
  <br/>
  <br/>
  <strong>PLAIN URL</strong>
  <br/>
  This is a simple redirect link for your Amazon listing without the Super URL. 
  <br/>
  Use this redirect to tag clicks with your retargeting pixel or simply to track click data.
  <br/>
  <br/>
  ⭐ ️<strong>REVIEW LINK</strong> ⭐️
  <br/>
  This link directs traffic to the "Create a Review" page, set with a default 5-star rating.
  <br/>
  <br/>
  <strong>ADD TO CART LANDING PAGE</strong>
  <br/>
  This generates an Add to Cart landing page with instructions.
  <br/>
  <br/>
  🚀 <strong>AMAZON CHOISE URL</strong> 🚀
  <br/>
  This creates search results based on "Alexa's Top Picks" with a specific keyword and ASIN.
  <br/>
  <br/>
  1️⃣ <strong>AMAZON BEST SELLER URL</strong> 1️⃣
  <br/>
  This creates search results based on "Alexa's Best Seller" with a specific keyword and ASIN.
  <br/>
  <br/>
  🔵 <strong>ALEXA LINK</strong> 🔵
  <br/>
  <br/>
  COMING SOON
  <br/>
  <br/>
  🔥 <strong>AFFILIATE LINK</strong> 🔥
  <br/>
  <br/>
  COMING SOON
  <br/>
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev.png">
  <br/>
  <br/>
</p>



<h2 id="made-in" align="center"><br/>🛠  Built in  🛠</h2>
  
<p align="center">
  The bot uses
  <a href="https://github.com/eternnoir/pyTelegramBotAPI">pyTelegramBotAPI</a>.
  <br/>
  <br/>
  More details can be found 
  <a href="#third-party-licenses">here</a>.
  <br/>
  <br/>
  Languages used:
  <br/>
  <br/>
  <a href="https://www.python.org">Python</a>
</p>

<p align="right"><a href="#top">⇧</a></p>



<h2 id="index" align="center"><br/>📋  Index  📋</h2>

<p align="center">
  <h4 align="center">
    <a href="#prerequisites">Prerequisites</a>
  </h4>
  <h4 align="center">
    <a href="#how-to-start">How to Start</a>
  </h4>
  <h4 align="center">
    <a href="#structure-of-the-project">Structure of the Project</a>
  </h4>
  <h4 align="center">
    <a href="#roadmap">Roadmap</a>
  </h4>
  <h4 align="center">
    <a href="#responsible-disclosure">Responsible Disclosure</a>
  </h4>
  <h4 align="center">
    <a href="#report-a-bug">Report a Bug</a>
  </h4>
  <h4 align="center">
    <a href="#maintenance">Maintenance</a>
  </h4>
  <h4 align="center">
    <a href="#license">License</a>
  </h4>
  <h4 align="center">
    <a href="#third-party-licenses">Third Party Licenses</a>
  </h4>
</p>

<p align="right"><a href="#top">⇧</a></p>



<h2 id="prerequisites" align="center"><br/>🧰  Prerequisites  🧰</h2>

<p align="center">
  Before starting you need to install the dependencies.
  <br/>
  In this case the only one needed is that of the pyTelegramBotAPI.
  <br/>
  <br/>
</p>
  
  
  ```
  pip3 install pyTelegramBotAPI
  ```
  
<p align="center">
  <br/>
  Once you have downloaded the program you need to create the bot.
  <br/>
  To do this, simply open the telegram app, search
  <br/>
  <strong>@BotFather</strong>
  <br/>
  Start by clicking on <strong>/start</strong> and continue to generate the new bot.
  <br/>
  You will need to choose your name and username.
  <br/>
  If you want, you can also upload your bot's profile image from here.
  <br/>
  <br/>
  Once generated, retrieve your bot Token.
  <br/>
  The Token is an alphanumeric string with symbols that uniquely identifies the bot.
  <br/>
  <br/>
  <strong>DO NOT SHARE THE TOKEN WITH ANYONE</strong>
  <br/>
  This token is strictly personal.
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev2.png">
  <br/>
  <br/>
  Now you need to recover your Telegram ID.
  <br/>
  To do this, look for the bot
  <br/>
  <strong>@userinfobot</strong>
  <br/>
  follow the instructions and recover your ID.
  <br/>
  (It is a sequence of 9 personal numbers that identifies your account)
  <br/>
  also in this case
  <br/>
  <br/>
  <strong>DO NOT SHARE YOUR TELEGRAM ID WITH ANYONE</strong>
  <br/>
  This is an example of token ID: 123123608
  <br/>
  <br/>
  You can finde more info
  <a href="https://www.alphr.com/telegram-find-user-id/"><strong>here</strong></a>
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev3.png">
  <br/>
</p>
  
<p align="right"><a href="#top">⇧</a></p>



<h2 id="how-to-start" align="center"><br/>⚙️  How to Start  ⚙️</h2>

<p align="center"> 
  <i>At this point you have everything you need to get started.</i>
  <br/>
  <i>Download the file and open it with your IDE or as you prefer.</i>
  <br/>
  <br/>
  Now we are going to enter your Telegram ID and the Token of your bot in the code.
  <br/>
  Your telegram ID is used to prevent third parties from using your bot without your consent.
  <br/>
  <br/>
  <br/>
  Open the <strong>main.py</strong> file
  <br/>
  At the beginning of the file you will have to replace the phrase <strong>YOUR-TELEGRAM-ID</strong> with your Telegram ID
  <br/>
  and then the phrase <strong>YOUR-TELEGRAM-API-TOKEN</strong> with your bot's Token.
  <br/>
  Once this is done, your bot is ready to go.
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev4.png">
  <br/>
  <br/>
  To start it open the terminal and positioning yourself in the same folder as the <strong>main.py</strong> file and type:
  <br/>
</p>

<p align="center">
  
  ```sh
  python3 main.py
  ```
</p>

<p align="center">
  <br/>
  Now you can generate your links <strong>FAST and FREE!</strong>
  <br/>
</p>


<p align="right"><a href="#top">⇧</a></p>



<h2 id="structure-of-the-project" align="center"><br/><br/>💻  Structure of the Project  💻</h2>

<p align="center">
  Using the bot is very simple.
  <br/>
  Open the telegram chat with your bot.
  <br/>
  You have 3 commands for information:
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev8.png">
  <br/>
  <br/>
  <strong>/list</strong>
  <br/>
  It tells you all the links you can choose
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev7.png">
  <br/>
  <br/>
  <strong>/spiegazione</strong>
  <br/>
  It gives you the explanation of all the links.
  <br/>
  <br/>
  So you can choose the most suitable link based on the result you want to achieve. Increase sales, rankings etc...
  <br/>
  <br/>
  <strong>/utilizzo</strong>
  <br/>
  It explains how to use the bot.
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev9.png">
  <br/>
  <br/>
  In summary once you have chosen the link, type the number of the link preceded by a slash <strong>'/1'</strong>
  <br/>
  Now enter the data you are asked for <strong>ONE PER LINE AND IN THE ORDER YOU ARE REQUESTED</strong>.
  <br/>
  <br/>
  <img src="https://github.com/Piero24/AmzUrlGen/blob/main/icon/prev6.png">
  <br/>
  <br/>
  <strong>Done!</strong> The bot has generated the link that you can use in your ADS or wherever you prefer.
  <br/>
</p>

<p align="right"><a href="#top">⇧</a></p>



---
  


<h2 id="roadmap" align="center"><br/><br/>🛫  Roadmap  🛬</h2>

- [x] Multi-language README
    - [x] English
    - [x] Italian
- [x] CANONICAL URL
- [x] REVIEW LINK
- [x] AMAZON BEST SELLER URL
- [ ] ALEXA LINK
- [ ] AFFILIATE LINK
- [ ] Other marketplace
    - [ ] US
    - [ ] UK
    - [ ] FR
    - [ ] DE
    - [ ] ES

See the [open issues](https://github.com/Piero24/AmzUrlGen/issues) for a full list of proposed features (and known issues).

<p align="right"><a href="#top">⇧</a></p>



<h2 id="responsible-disclosure" align="center"><br/><br/>📮  Responsible Disclosure  📮</h2>

<p align="center">
  We assume no responsibility for an improper use of this code and everything related to it.
  <br/>
  We do not assume any responsibility for damage caused to people and / or objects in the use of the code.
  <br/>
  <strong>By using this code even in a small part, the developers are declined from any responsibility.</strong>
  <br/>
  <br/>
  It is possible to have more information by viewing the following links:
  <br/>
  <a href="#code-of-conduct"><strong>Code of conduct</strong></a>
   • 
  <a href="#license"><strong>License</strong></a>
</p>



 <!-- https://www.contributor-covenant.org/ -->

<h3 id="code-of-conduct" align="center"><br/><br/>📜  Code of conduct  📜</h3>

<p align="center">
  As members, collaborators and leaders we are committed to performance participation in our community a harassment-free experience for all, regardless of age, size, visible or invisible disability, ethnicity, sexual characteristics, gender identity and expression, level of experience, education, socioeconomic status, nationality, appearance, race, caste, color, religion or identity and sexual orientation.
  <br/>
  We are committed to acting and interacting in ways that help create an open community, welcoming, diverse, inclusive and healthy.
  <br/>
  <br/>
  The complete version of the code of conduct can be viewed at the following link:
  <br/>
  <a href="https://github.com/Piero24/AmzUrlGen/blob/main/CODE-OF-CONDUCT.md"><strong>Code of Conduct »</strong></a>
</p>



<!--
***Template `ISSUE-TEMPLATE.md`
***https://github.com/stevemao/github-issue-templates
-->

<h2 id="report-a-bug"  align="center"><br/><br/>🐛  Bug and Feature 🐛</h2>

<p align="center">
  To <strong>report a bug</strong> or to request 
  <br/>
  the implementation of <strong>new features</strong>,
  <br/>
  it is strongly recommended to use the 
  <br/>
  <a href="https://github.com/Piero24/AmzUrlGen/issues"><strong>ISSUES tool from Github »</strong></a>
  <br/>
  <br/>
  Here you may already find the answer to the problem you have encountered,
  <br/>
  in case it has already happened to other people.
  <br/>
  Otherwise you can report the bugs found.
  <br/>
  <br/>
  <strong>!! ATTENTION !! </strong>
  <br/>
  </strong>To speed up the resolution of problems, </strong>
  <br/>
  <strong> it is recommended to answer all the questions </strong>
  <br/>
  <strong>present in the request phase in an exhaustive manner.</strong>
  <br/>
  <br/>
  (Even in the phase of requests for the implementation of new functions, 
  <br/>
  we ask you to better specify the reasons for the request 
  <br/>
  and what final result you want to obtain).
  <br/>
  <br/>
</p>


  
<h2 id="maintenance" align="center"><br/><br/>🔧  Maintenance  🔧</h2>

<p align="center">
  There are currently no parts of the code under maintenance.
  <br/>
  You can quickly check the 
  <a href="#top">status of the project</a>
   at the top of the page.
  <br/>
  <br/>
</p>

 <br/>

<h3 id="changelog" align="center"><br/>📟  Changelog  📟</h3>

<p align="center">
  Here you can find all the information regarding the fixes 
  <br/>
  and implementations that took place in the various program reviews.
  <br/>
  <br/>
  <a href="https://github.com/Piero24/AmzUrlGen/blob/main/CHANGELOG.md"><strong>Explore the changelog docs</strong></a>
  <br/>
</p>

<p align="right"><a href="#top">⇧</a></p>
  
  
  
 --- 
 
 
 
 <!--
<p align="center">
  <img src="https://cdn.dribbble.com/users/1292677/screenshots/6139167/media/fcf7fd0c619bb87706533079240915f3.gif" width="390" height="292">
</p>
 
-->
 
 
 
 <h2 id="license" align="center"><br/><br/><br/>🔍  License  🔍</h2>

<p align="center">
  The licensing section is divided into:
  <br/>
  <br/>
  - <a href="#general-license">General License</a>
  <br/>
  - <a href="#authors-and-copyright">Authors and Copyright</a>
  <br/>
  <br/>
  The 
  <a href="#general-license">General License</a> 
   indicates the base license chosen for the code.
  <br/>
  While 
  <a href="#authors-and-copyright">Authors and Copyright</a> 
   indicates who are the authors and who owns the copyright of the code.
  <br/>
</p>

<br/>

<!--
*** To choose the licenses you can look at the following links.
*** https://choosealicense.com/
*** https://github.com/teamdigitale/licenses
-->

<h3 id="general-license" align="center"><br/>🖋  General License  🖋</h3>

<p align="center">
  <strong>GNU GENERAL PUBLIC LICENSE</strong>
  <br/>
  <i>Version 3, 29 June 2007</i>
  <br/>
  <br/>
  <i>Copyright (C) 2007 Free Software Foundation, Inc. <http://fsf.org/></i>
  <br/>
  <i>Everyone is permitted to copy and distribute verbatim copies</i>
  <br/>
  <i>of this license document, but changing it is not allowed.</i>
  <br/>
  <br/>
  <i>Preamble</i>
  <br/>
  <i>The GNU General Public License is a free, copyleft license for</i>
  <br/>
  <i>software and other kinds of works.</i>
  <br/>
  <br/>
  <a href="https://github.com/Piero24/AmzUrlGen/blob/main/LICENSE"><strong>License Documentation »</strong></a>
  <br/>
</p>

<br/>

<!--
<p align="center">
  <img src="https://cdn.dribbble.com/users/1292677/screenshots/6139167/media/fcf7fd0c619bb87706533079240915f3.gif" width="390" height="292">
</p>
 
-->

<h3 id="authors-and-copyright" align="center"><br/>✏️  Authors and Copyright  ✏️</h3>

<p align="center">
  👨🏽‍💻: <strong>Pietrobon Andrea</strong>
  <br/>
  🌐: 
  <a href="https://www.pietrobonandrea.com">www.pietrobonandrea.com</a>
  <br/>
  <img src="https://assets.stickpng.com/thumbs/580b57fcd9996e24bc43c53e.png" width="30" height="30" align="center">:
  <a href="https://twitter.com/pietrobonandrea">@PietrobonAndrea</a>
  <br/>
  🗄:
  <a href="https://github.com/Piero24/AmzUrlGen">AmzUrlGen</a>
  <br/>
  <br/>
  <br/>
  My name is <strong>Pietrobon Andrea</strong>,
  <br/>
  a computer engineering student at the 
  <br/>
  <img src="https://upload.wikimedia.org/wikipedia/it/thumb/5/53/Logo_Università_Padova.svg/800px-Logo_Università_Padova.svg.png"  width="30" height="30" align="center"> University of Padua (🇮🇹).
  <br/>
  <br/>
  My passion turns towards <strong>AI</strong> and <strong>ML</strong>.
  <br/>
  I have learned and worked in different sectors that have allowed me to gain skills 
  <br/>
  in different fields, such as IT and industrial design.
  <br/>
  <br/>
  To find out more, visit my 
  <br/>
  <a href="https://www.pietrobonandrea.com"><strong>website »</strong></a>
  <br/>
  <br/>
  The license of this repo as indicated in the 
  <a href="#general-license">general license</a> section is the 
  <br/>
  GNU General Public License v3.0
  <br/>
  For more details visit the 
  <a href="https://github.com/Piero24/AmzUrlGen/blob/main/LICENSE"><strong>licensing page »</strong></a>
  <br/>
  <br/>
  <strong>The Copyright (C) of this project is held exclusively by my person.</strong>
</p>

<p align="right"><a href="#top">⇧</a></p>


<!--
 ***This operation could be particularly long or laborious to report and therefore it is possible to avail
 *** of some automatic support tools.
 ***https://fossa.com/pricing/
-->

<h2 id="third-party-licenses" align="center"><br/>📌  Third Party Licenses  📌</h2>

<p align="center">
  In the event that the software uses third-party components for its operation, 
  <br/>
  the individual licenses are indicated in the following section.
  <br/>
  <br/>
  <strong>Software list:</strong>
  <br/>
  <table align="center">
  <tr  align="center">
    <th>Software</th>
    <th>License owner</th> 
    <th>License type</th> 
    <th>Link</th>
    <th>Note</th>
  </tr>
  <tr  align="center">
    <td>pyTelegramBotAPI</td>
    <td><a href="https://github.com/eternnoir">eternoir</a></td>
    <td><a href="https://github.com/eternnoir/pyTelegramBotAPI/blob/master/LICENSE">GNU GPL v2.0</a></td>
    <td><a href="https://github.com/eternnoir/pyTelegramBotAPI">github</a></td>
    <td></td>
  </tr>
</table>
</p>


<p align="right"><a href="#top">⇧</a></p>



---
<br/>

> *<p align="center"> Copyrright (C) by Pietrobon Andrea <br/> Released date: May-20-2022*


CHANGELOG.md
# Changelog



## v1.1.0

### Added or Changed
- Added CANONICAL URL
- Added REVIEW LINK
- Added AMAZON BEST SELLER URL
- Fix some bug
- Changed table of contents to start collapsed

### Removed

- Some packages/libraries from acknowledgements I no longer use


back_url.py

def link_choise(final_link, stat_var, list_link):

    if stat_var in ['1']:

        final_link += f"{list_link[1].replace(' ', '+')}/s?k={list_link[1].replace(' ', '+')}&" \
                      f"rh=p_78%3{list_link[0].replace(' ', '')}&" \
                      f"maas=maas_adg_1&ref=aa_maask%3D{list_link[1].replace(' ', '+')}"

    elif stat_var in ['2']:

        final_link += f"s?k={list_link[1].replace(' ', '+')}&dc&hidden-keywords={list_link[0].replace(' ', '')}"

    elif stat_var in ['3']:

        final_link += f"fdp/{list_link[0].replace(' ', '')}/ref=sr_1_2?ie=UTF8&" \
                      f"keywords={list_link[1].replace(' ', '+')}&qid=1637195997&s=gateway&sr=8-2"

    elif stat_var in ['4']:

        final_link += f"s/ref=nb_sb_noss_1?url=search-alias%3Daps&field-keywords={list_link[0].replace(' ', '+')}"

    elif stat_var in ['5']:

        final_link += f"s?k={list_link[1].replace(' ', '+')}&" \
                      f"rh=p_4%3A{list_link[2].replace(' ', '%20')}%2Cp_78" \
                      f"%3A{list_link[0].replace(' ', '')}%2Cp_36%3A1500-2500%2Cssx%3Arelevance&ref=nb_sb_noss_2"

    elif stat_var in ['6']:

        final_link += "gp/aws/cart/add.html?"

        for i in range(1, int(len(list_link) / 2)):
            final_link += f"ASIN.{i}=" + list_link[2 * i].replace(' ', '') + \
                          f"&Quantity.{i}=" + list_link[2 * i + 1].replace(' ', '') + "&"

        final_link = final_link[:-1]

    elif stat_var in ['7']:

        final_link += f"gp/aws/cart/add.html?ASIN.1={list_link[0].replace(' ', '')}&" \
                      f"Quantity.1={list_link[1].replace(' ', '')}"

    elif stat_var in ['8']:

        for i in range(1, len(list_link)):
            final_link += f"{list_link[i].replace(' ', '+')}-"

        final_link = final_link[:-1]
        final_link += f"/dp/{list_link[0].replace(' ', '')}/?maas=ampattrib"

    elif stat_var in ['9']:

        final_link += f"dp/*ASIN{list_link[0].replace(' ', '')}*?"

    elif stat_var in ['10']:

        final_link += f"review/create-review?&asin={list_link[0].replace(' ', '')}:5"

    elif stat_var in ['11']:

        final_link += f"gp/aws/cart/add.html?ASIN.1={list_link[0].replace(' ', '')}&" \
                      f"Quantity.1={list_link[1].replace(' ', '')}"

    elif stat_var in ['12']:

        final_link += f"search?ref=msv_sr&keyword={list_link[1].replace(' ', '+')}&" \
                      f"asins={list_link[0].replace(' ', '')}-C"

    elif stat_var in ['13']:

        final_link += f"search?ref=msv_sr&keyword={list_link[1].replace(' ', '+')}&" \
                      f"asins={list_link[0].replace(' ', '')}-B"

    elif stat_var in ['14']:

        final_link += f"gp/aws/cart/add.html?&ASIN.1={list_link[0].replace(' ', '')}&" \
                      f"Quantity.1={list_link[2].replace(' ', '')}&keywords={list_link[1].replace(' ', '+')}"

    elif stat_var in ['15']:

        pass

    return final_link

main.py
# pip3 install pyTelegramBotAPI
import time
import telebot
import list_file as lf

id_list = ["YOUR-TELEGRAM-ID"]
API_TOKEN = 'YOUR-TELEGRAM-API-TOKEN'
bot = telebot.TeleBot(API_TOKEN)
stat_var = 0

@bot.message_handler(commands=lf.command_list)
def send_welcome(message):

    global stat_var
    arrived_msg = message.text.replace('/', '')

    # Effettua il typing a schermo per l'utente
    bot.send_chat_action(message.from_user.id, "typing")
    time.sleep(1)

    if message.from_user.id in id_list:

        if arrived_msg == 'list': bot.send_message(message.from_user.id, lf.str_list, parse_mode='MarkdownV2')

        elif arrived_msg == 'spiegazione': bot.send_message(message.from_user.id, lf.what_the_link_do, parse_mode='MarkdownV2')

        elif arrived_msg == 'utilizzo': bot.send_message(message.from_user.id, lf.how_to_use, parse_mode='MarkdownV2')

        elif arrived_msg in lf.command_list and arrived_msg != 'list' \
                and arrived_msg != 'spiegazione' and arrived_msg != 'utilizzo':

            if arrived_msg in ['1', '2', '3', '12', '13']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN e una KEYWORD")
            elif arrived_msg in ['4']:
                bot.send_message(message.from_user.id, "Inserisci una KEYWORD")
            elif arrived_msg in ['5']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN una KEYWORD e il nome del BRAND")
            elif arrived_msg in ['6']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN e una quantità per ogni prodotto (da 2 a 4 prodotti)")
            elif arrived_msg in ['7', '11']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN e la quantità di unità")
            elif arrived_msg in ['8']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN e da 2 a 5 KEYWORD")
            elif arrived_msg in ['9', '10']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN")
            elif arrived_msg in ['14']:
                bot.send_message(message.from_user.id, "Inserisci un ASIN una KEYWORD e la quantità di unità")

            stat_var = arrived_msg

    else:

        bot.send_message(message.from_user.id, "Accesso negato.")
        bot.send_chat_action(message.from_user.id, "typing")
        time.sleep(1)
        bot.send_message(message.from_user.id, "Il tuo tentativo di accesso è stato reportato.")
        bot.send_message(id_list[0], "C'è stato un tentativo di accesso non autorizzato.")
        bot.send_message(id_list[0], "ID: " + message.from_user.id + "\nTEXT: " + message.text)


@bot.message_handler(func=lambda message: True)
def echo_message(message):

    
    global stat_var
    bot.send_chat_action(message.from_user.id, "typing")
    time.sleep(1)

    if message.from_user.id in id_list:

        if message.text == 'exit': stat_var = 0

        elif stat_var == 0: bot.send_message(id_list[0], "Seleziona prima un tipo di URL")

        elif stat_var != 0:

            list_link = message.text.split('\n')
            final_link = f"https://www.amazon.it/{lf.link_list_el[int(stat_var) - 1]}"

            if stat_var in ['1', '2', '3', '12', '13']:

                try:
                    final_link = final_link.replace("*ASIN*", list_link[0].replace(' ', ''))
                    final_link = final_link.replace("*KEYWORD*", list_link[1].replace(' ', '+'))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['4']:

                try:
                    final_link = final_link.replace("*KEYWORD*", list_link[0].replace(' ', '+'))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['5']:

                try:
                    final_link = final_link.replace("*ASIN*", list_link[0].replace(' ', ''))
                    final_link = final_link.replace("*KEYWORD*", list_link[1].replace(' ', '+'))
                    final_link = final_link.replace("*BRAND*", list_link[2].replace(' ', '%20'))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['6']:

                # verificare se viene inserita la prima parte del link "gp/aws/cart/add.html?"

                try:
                    for i in range(1, int(len(list_link)/2)):
                        final_link += f"ASIN.{i}=" + list_link[2*i].replace(' ', '') + \
                        f"&Quantity.{i}=" + list_link[2*i + 1].replace(' ', '') + "&"

                    final_link = final_link[:-1]
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['7', '11']:

                try:
                    final_link = final_link.replace("*ASIN*", list_link[0].replace(' ', ''))
                    final_link = final_link.replace("*QUANTITY*", list_link[1].replace(' ', ''))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['8']:

                try:
                    for i in range(1, len(list_link)):
                        final_link += f"{list_link[i].replace(' ', '+')}-"

                    final_link = final_link[:-1]
                    final_link += f"/dp/{list_link[0].replace(' ', '')}/?maas=ampattrib"

                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")


            elif stat_var in ['9', '10']:

                try:
                    final_link = final_link.replace("*ASIN*", list_link[0].replace(' ', ''))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            elif stat_var in ['14']:

                try:
                    final_link = final_link.replace("*ASIN*", list_link[0].replace(' ', ''))
                    final_link = final_link.replace("*KEYWORD*", list_link[1].replace(' ', '+'))
                    final_link = final_link.replace("*QUANTITY*", list_link[2].replace(' ', ''))
                except:
                    bot.send_message(message.from_user.id, "Devi aver inserito i dati in modo errato.\nRiprova.")

            bot.send_message(message.from_user.id, final_link)
            stat_var = 0

    else:

        bot.send_message(message.from_user.id, "Accesso negato.")
        bot.send_chat_action(message.from_user.id, "typing")
        time.sleep(1)
        bot.send_message(message.from_user.id, "Il tuo tentativo di accesso è stato reportato.")
        bot.send_message(id_list[0], "C'è stato un tentativo di accesso non autorizzato.")
        bot.send_message(id_list[0], "ID: " + message.from_user.id + "\nTEXT: " + message.text)
    


bot.infinity_polling(timeout=10, long_polling_timeout = 5)